crates/update/src/cli/install.rs

```diff
@@ -0,0 +1,207 @@
+    let url = match &version {
+        Some(version) => format!("{releases_url}/tags/v{version}"),
+        None => [releases_url, "/latest"].concat(),
+    };
+    let release: Release = client
+        .get(url)
+        .send()
+        .await?
+        .error_for_status()
+        .map_err(|e| {
+            if e.status() == Some(reqwest::StatusCode::NOT_FOUND) {
+                if let Some(version) = &version {
+                    return anyhow::anyhow!(e).context(format!("No release found for version {version}"));
+                }
+            }
+            anyhow::anyhow!(e).context("Could not fetch release info")
+        })?
+        .json()
+        .await?;
+    let release_version = match version {
+        Some(version) => version,
+        None => release.version().context("Could not parse version number")?,
+    };
+
+    let asset = release
+        .assets
+        .iter()
+        .find(|&asset| asset.name == download_name)
+        .ok_or_else(|| {
+            let err = anyhow::anyhow!("artifact named {download_name} not found in version {release_version}");
+            if custom_artifact {
+                err
+            } else {
+                err.context("no prebuilt binaries available for the detected OS and architecture")
+            }
+        })?;
+
+    pb.set_style(ProgressStyle::with_template("{spinner} {prefix}{msg} {bytes}/{total_bytes} ({eta})").unwrap());
+    pb.set_prefix(format!("Installing v{release_version}: "));
+    pb.set_message("downloading...");
+    let archive = download_with_progress(&pb, client, &asset.browser_download_url).await?;
+
+    pb.set_style(pb_style);
+    pb.set_message("unpacking...");
+
+    let version_dir = paths.cli_bin_dir.version_dir(&release_version);
+    match artifact_type {
+        ArtifactType::TarGz => {
+            let tgz = archive.aggregate().reader();
+            tar::Archive::new(flate2::bufread::GzDecoder::new(tgz)).unpack(&version_dir)?;
+        }
+        ArtifactType::Zip => {
+            let zip = archive.to_bytes();
+            let zip = io::Cursor::new(&*zip);
+            zip::ZipArchive::new(zip)?.extract(&version_dir)?;
+        }
+    }
+
+    pb.finish_with_message("done!");
+
+    Ok(release_version)
+}
+
+enum ArtifactType {
+    TarGz,
+    Zip,
+}
+
+impl ArtifactType {
+    fn deduce(filename: &str) -> Option<Self> {
+        if filename.ends_with(".tar.gz") {
+            Some(Self::TarGz)
+        } else if filename.ends_with(".zip") {
+            Some(Self::Zip)
+        } else {
+            None
+        }
+    }
+}
+
+pub(super) async fn available_releases(client: &reqwest::Client) -> anyhow::Result<Vec<semver::Version>> {
+    let url = "https://api.github.com/repos/clockworklabs/SpacetimeDB/releases";
+    let releases: Vec<Release> = client.get(url).send().await?.json().await?;
+
+    releases.into_iter().map(|release| release.version()).collect()
+}
+
+#[derive(Deserialize)]
+struct ReleaseAsset {
```

Comment:
can we rename these to indicate that they're API types (not types that we get to choose).

e.g.
```suggestion
// The type returned by GitHub's API to describe a release asset
struct GHApiReleaseAsset {
```

Reply:
nevermind, I realize these were transplanted from the previous code :+1: can always clean it up later